File: ConvertMeterReading/src/main/java/com/crapp/smartmeter/ProcessData.java
package com.crapp.smartmeter;

import gnu.io.CommPort;
import gnu.io.CommPortIdentifier;
import gnu.io.PortInUseException;
import gnu.io.SerialPort;

import java.io.*;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.Scanner;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;


/**
 * Created by efo20922 on 4-1-2016.
 */
public class ProcessData {

    MeasurementsDataBase store = MeasurementsDataBase.getInstance();

    public ProcessData() {
        System.out.println("ProcessData");
    }

    public void connect(String portName) throws Exception {
        System.out.println("connect on usb" + portName);

        CommPortIdentifier portIdentifier = CommPortIdentifier.getPortIdentifier(portName);

        if (portIdentifier.isCurrentlyOwned()) {
            System.out.println("Error: Port is currently in use");
        } else {

            int timeout = 2000;
            while (1 != 0) {

                try {
                    CommPort commPort = portIdentifier.open(this.getClass().getName(), timeout);
                    if (commPort instanceof SerialPort) {
                        SerialPort serialPort = (SerialPort) commPort;

                        /*
                        The interface will use a fixed transfer speed of 115200 baud. For character formatting a start
                        bit, 8 data bits, no parity bit and a stop bit are used (8N1).Note this is not conforming to
                        ENIEC62056-21 Mode D.
                         */
                        serialPort.setSerialPortParams(115200,
                                SerialPort.DATABITS_8,
                                SerialPort.STOPBITS_1,
                                SerialPort.PARITY_NONE);

                        InputStream in = serialPort.getInputStream();

                        Calendar calendar = Calendar.getInstance();
                        ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();
                        scheduler.scheduleAtFixedRate(new SerialReader(in, commPort),
                                millisToNextHour(calendar), 60 * 60 * 1000, TimeUnit.MILLISECONDS);

                        //Thread myThread = new Thread(new SerialReader(in, commPort));
                        //myThread.start();
                    } else {
                        System.out.println("Error: Only serial ports are handled by this example.");
                    }
                } catch (PortInUseException ex) {
                    System.out.println("PortInUseException , lets wait for another turn");
                }

                Thread.sleep(1800000);
            }
        }
    }

    private static long millisToNextHour(Calendar calendar) {
        int hour = calendar.get(Calendar.HOUR_OF_DAY);
        int minutes = calendar.get(Calendar.MINUTE);
        int seconds = calendar.get(Calendar.SECOND);
        int millis = calendar.get(Calendar.MILLISECOND);
        int minutesToNextHour = 60 - minutes - 1;
        int secondsToNextHour = 60 - seconds - 1;
        int millisToNextHour = 1000 - millis;

        System.out.println("Now: " + hour + ":" +minutes + ":" + seconds);
        System.out.println("Fire in: " + minutesToNextHour + ":" + secondsToNextHour + " minutes");

        return minutesToNextHour*60*1000 + secondsToNextHour*1000 + millisToNextHour;
    }

    private class SerialReader implements Runnable {

        InputStream in;
        CommPort commPort;

        public SerialReader(InputStream in, CommPort commPort) {
            this.in = in;
            this.commPort = commPort;
        }

        public void run() {
            System.out.println("start new thread on " + new Date());
            BufferedReader br = new BufferedReader(new InputStreamReader(in));
            String line;
            StringBuffer buf = new StringBuffer();

            // continue processing
            try {
                while ((line = br.readLine()) != null) {
                    buf.append(line).append(System.lineSeparator());

                    if (line.startsWith("!")) {

                        processSmartMeter(buf.toString());

                        in.close();
                        commPort.close();
                        System.out.println("return thread");
                        return;
                    }
                }
            } catch (Exception ex) {
                System.out.println(ex.toString());
                Thread.currentThread().interrupt();
            }
        }
    }

    /*
    Exceptions
    java.lang.NumberFormatException: For input string: "001477.458*?1-0:2.8.1(000000.000"
    str = str.replaceAll("\\D+","");
    1-0:1.8.1(001434.912*kWh)
    1-0:2.8.1(000000.000*kWh)
    */
    public void processSmartMeter(String result) throws Exception {

        Scanner scanner = new Scanner(result);

        // --== ONLY USE FOR TEST START ==--
        // FileReader fileReader = new FileReader(new File("messages_long.txt"));
        // BufferedReader br = new BufferedReader(fileReader);
        // --== ONLY USE FOR TEST END   ==--

        Date dateElectric = new Date();
        String deviceElectric = "";
        Float meter181 = 0f;
        Float meter281 = 0f;
        Float meter182 = 0f;
        Float meter282 = 0f;
        Integer tarif = 0;
        Float powerDeliveredKw = 0f;
        Float powerReceivedKw = 0f;
        Integer powerFailures = 0;
        Integer longPowerFailures = 0;
        String deviceGas = "";
        Date dateGas = new Date();
        Float gasMeasurement = 0f;

        // String line;
        // while ((line = br.readLine()) != null) {
            while (scanner.hasNextLine()) { // for testing

            String replace = "[\\s()*kWh?]";
            try {
                String line = scanner.nextLine().trim();
                if (line.startsWith("0-0:1.0.0")) {
                    line = line.replace("0-0:1.0.0", "");
                    line = line.replaceAll(replace, "");
                    dateElectric = new SimpleDateFormat("yyMMddhhmmss", Locale.ENGLISH).parse(line);

                    // Equipment identifier
                } else if (line.startsWith("0-0:96.1.1")) {
                    line = line.replace("0-0:96.1.1", "");
                    line = line.replaceAll(replace, "");
                    deviceElectric = line.substring(0, 34);

                    // Meter Reading electricity delivered to client (Tariff 1)
                } else if (line.startsWith("1-0:1.8.1")) {
                    line = line.replace("1-0:1.8.1", "");
                    line = line.replaceAll(replace, "");
                    meter181 = Float.valueOf(line.substring(0, 10));

                } else if (line.startsWith("1-0:2.8.1")) {
                    line = line.replace("1-0:2.8.1", "");
                    line = line.replaceAll(replace, "");
                    meter281 = Float.valueOf(line.substring(0, 10));

                    // Meter Reading electricity delivered by client (Tariff 2)
                } else if (line.startsWith("1-0:1.8.2")) {
                    line = line.replace("1-0:1.8.2", "");
                    line = line.replaceAll(replace, "");
                    meter182 = Float.valueOf(line.substring(0, 10));

                    // Meter Reading electricity delivered by client (Tariff 2)
                } else if (line.startsWith("1-0:2.8.2")) {
                    line = line.replace("1-0:2.8.2", "");
                    line = line.replaceAll(replace, "");
                    meter282 = Float.valueOf(line.substring(0, 10));

                    //Tariff indicator electricity. The tariff indicator can be used to switch tariff dependent loads e.g boilers.
                    // This is responsibility of the P1 user
                } else if (line.startsWith("0-0:96.14.0")) {
                    line = line.replace("0-0:96.14.0", "");
                    line = line.replaceAll(replace, "");
                    tarif = Integer.valueOf(line.substring(0, 4));

                    // Actual electricity power delivered (+P) in 1 Watt resolution
                } else if (line.startsWith("1-0:1.7.0")) {
                    line = line.replace("1-0:1.7.0", "");
                    line = line.replaceAll(replace, "");
                    powerDeliveredKw = (Float.valueOf(line.substring(0, 6)) * 1000);

                    // Actual electricity power received (-P) in 1 Watt resolution
                } else if (line.startsWith("1-0:2.7.0")) {
                    line = line.replace("1-0:2.7.0", "");
                    line = line.replaceAll(replace, "");
                    powerReceivedKw = (Float.valueOf(line.substring(0, 6)) * 1000);

                    // Number of power failures in any phases
                } else if (line.startsWith("0-0:96.7.21")) {
                    line = line.replace("0-0:96.7.21", "");
                    line = line.replaceAll(replace, "");
                    powerFailures = Integer.valueOf(line.substring(0, 5));

                    // Number of long power failures in any phases
                } else if (line.startsWith("0-0:96.7.9")) {
                    line = line.replace("0-0:96.7.9", "");
                    line = line.replaceAll(replace, "");
                    longPowerFailures = Integer.valueOf(line.substring(0, 5));

                    // Equipment identifier
                } else if (line.startsWith("0-1:96.1.0")) {
                    line = line.replace("0-1:96.1.0", "");
                    line = line.replaceAll(replace, "");
                    deviceGas = line.substring(0, 34);

                    // Gas measurement
                } else if (line.startsWith("0-1:24.2.1")) {
                    line = line.replace("0-1:24.2.1", "");
                    line = line.replaceAll(replace, "");

                    String dateString = line.substring(0, 12);
                    dateGas = new SimpleDateFormat("yyMMddhhmmss", Locale.ENGLISH).parse(dateString);
                    line = line.replace(dateString, "");

                    gasMeasurement = Float.valueOf(line.substring(0,9));
                }

            } catch (NumberFormatException nfe) {
                System.out.println(nfe.toString() + "\n");
                nfe.printStackTrace();
            } catch (ParseException pe) {
                System.out.println(pe.toString() + "\n");
                pe.printStackTrace();
            } catch (Exception e) {
                System.out.println(e.toString() + "\n");
                e.printStackTrace();
            }
        }

        //        System.out.println(
        //                "Date Electric: " + dateElectric + "\n" +
        //                        "Device ID Electric: " + deviceElectric + "\n" +
        //                        "Meter181: " + meter181 + "\n" +
        //                        "Meter281: " + meter281 + "\n" +
        //                        "Meter182: " + meter182 + "\n" +
        //                        "Meter281: " + meter282 + "\n" +
        //                        "Tarif: " + tarif + "\n" +
        //                        "Power Deliverd: " + powerDeliveredKw + "\n" +
        //                        "Power Received: " + powerReceivedKw + "\n" +
        //                        "Power Failures: " + powerFailures + "\n" +
        //                        "Long Power Failures: " + longPowerFailures + "\n" +
        //                        "Device ID Gas: " + deviceGas + "\n" +
        //                        "Date Gas: " + dateGas + "\n" +
        //                        "Gas Measurement: " + gasMeasurement);

        store.addEnergyMeasurement(
                dateElectric, deviceElectric,
                meter181, meter281, meter182, meter282,
                tarif, powerDeliveredKw, powerReceivedKw,
                powerFailures, longPowerFailures, deviceGas, dateGas,
                gasMeasurement);


    }

    public static void main(String[] args) {

        try {
            (new ProcessData()).connect("/dev/ttyUSB1");
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: nodeapp/processdata.js
var fs = require('fs');
var serialport = require('serialport');
var SerialPort = serialport.SerialPort;
var mongoose = require('mongoose');
var active = false;
var message = [];

/* 
 * This function is exported
 */
function openSerialPort() {

    console.log('Opening serialport... at ' + new Date());
    
    try {
        if (!active) {
            active = true;
            
            var serialPort = new SerialPort('/dev/ttyUSB0', {
                baudrate: 115200,
                dataBits: 8,
                stopBits: 1,
                parity: 'none',
                parser: serialport.parsers.readline('\n')
            });

            /* 
             * Reads the serial port per line.
             */
            serialPort.on("data", function(data) {
                data = data.replace(/\u0000/g, '');
                data = data.replace(/\r/g, '');

                // if(startsWith(data, '/')){
                //     console.log('startsWith /');
                // }

                // if (data == "" || data == " ") { return; }
                console.log('Data: ' + data);

                if (data !== '') {
                    message.push(data);
                }

                if (data.charAt(0) === '!') {
                    console.log('! found in message');

                    // console.log(message);

                    if (message.length != 25) {
                        console.log('length not 25, but: ' + message.length);
                        message = [];
                        console.log('Is there something left?' + data);
                        // serialPort.flush(function(err,results){});
                        // data = [];
                        console.log('Lenght now: ' + message.length);
                    } else {
                        // convert message to something usefull
                        var msg = obtainMeasurement(message);

                        // load mongoose schema
                        var Measurement = mongoose.model('measurement');
                        var myMeasurement = new Measurement(msg);

                        // Actual save to db.measurements
                        myMeasurement.save(function(err) {
                            if (err){
                               return handleError(err); 
                            } 
                            console.log('msg stored in db' + new Date());
                        });

                        message = [];

                        // close port and sleep for any given time
                        serialPort.close(function(err) {
                            console.log('port closed');
                            if (err) {
                                console.log('Error: ' + err);
                            }
                            active = false;
                        });
                    }
                }
            });
        }
    } catch (e) {
        // Error means port is not available for listening.
        console.log('something went wrong with the serialport');
        if(e){
            console.log(e);
        }
    }
}

/**
 * Expose methods to other files
 */
module.exports = {
    openSerialPort: function() {
        openSerialPort();
    }
}

/*
 *   Helper functions
 */
function convertToTimeStamp(date) {
    // Dates are formatted in YYMMDDHHMMSS 
    // Example 160301083129
    return new Date(2000 + +date.substr(0, 2),
        date.substr(2, 2) - 1,
        date.substr(4, 2),
        date.substr(6, 2),
        date.substr(8, 2),
        date.substr(10, 2)).getTime();
}

function replaceAll(str, find, replace) {
    return str.replace(new RegExp(find, 'g'), replace);
}

function startsWith(str, prefix) {
    if (str.length < prefix.length){
        return false;
    }
    for (var i = prefix.length - 1;
        (i >= 0) && (str[i] === prefix[i]); --i){
        continue; 
    }

    return i < 0;
}

function obtainMeasurement(message) {

    var dateElectric;
    var deviceElectric;
    var meter181kWh;
    var meter281kWh;
    var meter182kWh;
    var meter282kWh;
    var tariff;
    var powerDeliveredKw;
    var powerReceivedKw;
    var powerFailures;
    var powerFailuresLong;
    var deviceGas;
    var dateGas;
    var gasMeasurementm3;

    var replace = '[\\s()*kWh?]';

    var convertedMessage = [];

    for (var i = 0; i < message.length; i++) {
        if (startsWith(message[i], '0-0:1.0.0')) {
            // Date
            message[i] = message[i].replace('0-0:1.0.0', '');
            message[i] = replaceAll(message[i], replace, '');

            dateElectric = convertToTimeStamp(message[i]);

            convertedMessage['dateElectric'] = dateElectric;

        } else if (startsWith(message[i], '0-0:96.1.1')) {
            // Equipment identifier
            message[i] = message[i].replace('0-0:96.1.1', '');
            message[i] = replaceAll(message[i], replace, '');
            deviceElectric = message[i];

            convertedMessage['deviceElectric'] = deviceElectric;

        } else if (startsWith(message[i], '1-0:1.8.1')) {
            // Meter Reading electricity delivered to client (Tariff 1)
            message[i] = message[i].replace('1-0:1.8.1', '');
            message[i] = replaceAll(message[i], replace, '');
            meter181kWh = message[i];

            convertedMessage['meter181kWh'] = meter181kWh;

        } else if (startsWith(message[i], '1-0:2.8.1')) {
            // Meter Reading electricity delivered by client (Tariff 1)
            message[i] = message[i].replace('1-0:2.8.1', '');
            message[i] = replaceAll(message[i], replace, '');
            meter281kWh = message[i];

            convertedMessage['meter281kWh'] = meter281kWh;

        } else if (startsWith(message[i], '1-0:1.8.2')) {
            // Meter Reading electricity delivered to client (Tariff 2)
            message[i] = message[i].replace('1-0:1.8.2', '');
            message[i] = replaceAll(message[i], replace, '');
            meter182kWh = message[i];

            convertedMessage['meter182kWh'] = meter182kWh;

        } else if (startsWith(message[i], '1-0:2.8.2')) {
            // Meter Reading electricity delivered by client (Tariff 2)
            message[i] = message[i].replace('1-0:2.8.2', '');
            message[i] = replaceAll(message[i], replace, '');
            meter282kWh = message[i];

            convertedMessage['meter282kWh'] = meter282kWh;

        } else if (startsWith(message[i], '0-0:96.14.0')) {
            // Tariff indicator electricity. The tariff indicator can be used to switch tariff dependent loads e.g boilers.
            // This is reserialPortonsibility of the P1 user
            message[i] = message[i].replace('0-0:96.14.0', '');
            message[i] = replaceAll(message[i], replace, '');
            tariff = message[i];

            convertedMessage['tariff'] = Number(tariff);

        } else if (startsWith(message[i], '1-0:1.7.0')) {
            // Actual electricity power delivered (+P) in 1 Watt resolution
            message[i] = message[i].replace('1-0:1.7.0', '');
            message[i] = replaceAll(message[i], replace, '');
            powerDeliveredKw = message[i];

            convertedMessage['powerDeliveredKw'] = powerDeliveredKw;

        } else if (startsWith(message[i], '1-0:2.7.0')) {
            // Actual electricity power received (-P) in 1 Watt resolution
            message[i] = message[i].replace('1-0:2.7.0', '');
            message[i] = replaceAll(message[i], replace, '');
            powerReceivedKw = message[i];

            convertedMessage['powerReceivedKw'] = powerReceivedKw;

        } else if (startsWith(message[i], '0-0:96.7.21')) {
            // Number of power failures in any phases
            message[i] = message[i].replace('0-0:96.7.21', '');
            message[i] = replaceAll(message[i], replace, '');
            powerFailures = message[i];

            convertedMessage['powerFailures'] = Number(powerFailures);

        } else if (startsWith(message[i], '0-0:96.7.9')) {
            // Number of long power failures in any phases
            message[i] = message[i].replace('0-0:96.7.9', '');
            message[i] = replaceAll(message[i], replace, '');
            powerFailuresLong = message[i];

            convertedMessage['powerFailuresLong'] = Number(powerFailuresLong);

        } else if (startsWith(message[i], '0-1:96.1.0')) {
            // Equipment identifier
            message[i] = message[i].replace('0-1:96.1.0', '');
            message[i] = replaceAll(message[i], replace, '');
            deviceGas = message[i];

            convertedMessage['deviceGas'] = deviceGas;

        } else if (startsWith(message[i], '0-1:24.2.1')) {
            // Gas measurement
            message[i] = message[i].replace('0-1:24.2.1', '');
            message[i] = replaceAll(message[i], replace, '');

            var dateString = message[i].substring(0, 12);
            dateGas = convertToTimeStamp(dateString);

            message[i] = message[i].replace(dateString, '');
            gasMeasurementm3 = message[i].substring(0, 9);

            convertedMessage['dateGas'] = dateGas;
            convertedMessage['gasMeasurementm3'] = gasMeasurementm3;
            convertedMessage['date'] = new Date().setHours(0,0,0,0);
            convertedMessage['dateTime'] = new Date();
        }
    }
    return convertedMessage;
}